torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2003,6 +2014,300 @@ def clip_grad_norm_(
                 assert p.grad is not None
                 p.grad.detach().mul_(clip_coef.to(p.grad.device))
 
+    @staticmethod
+    def full_optim_state_dict(
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        optim_input: Optional[Union[
+            List[Dict[str, Any]], Iterable[torch.nn.Parameter],
+        ]] = None,
+    ) -> Dict[str, Any]:
+        """
+        Consolidates the full optimizer state on rank 0 and returns it
+        as a :class:`dict` following the convention of
+        :meth:`torch.optim.Optimizer.state_dict`, i.e. with keys ``"state"``
+        and ``"param_groups"``. The flattened parameters in ``FSDP`` modules
+        contained in ``model`` are mapped back to their unflattened parameters.
+
+        .. warning:: This needs to be called on all ranks since synchronization
+            primitives are used.
+
+        .. warning:: The parameter IDs used in the state dict depend on the
+        ``FSDP`` wrapping scheme used in the passed-in ``model``. This means
+        that they may not match those of a purely non-wrapped version of the
+        model and that the returned state dict cannot necessarily be loaded
+        for a version of the model with a different ``FSDP`` wrapping scheme.
+        However, this does support changing the world size between save and
+        load time.
+
+        .. warning:: If you do not pass ``model.parameters()`` as the first
+        argument to the optimizer, then you should pass that same value to
+        this method as ``optim_input``.
```

Comment:
curious, when the optimizer won't be initialized with model parameters?

Reply:
Multiple parameter groups. Example 1: if you want to apply weight decay to some parameters but not others. Example 2: if you want to train some layers with different learning rate.

Mathematically, once you allow multiple parameter groups, you can arbitrarily permute the order of the parameters passed into the optimizer constructor. This makes following the same unflattened parameter IDs as the purely non-wrapped equivalent challenging.